src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,147 @@
 package core.basesyntax;
 
 import java.util.List;
+import java.util.Objects;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private class Node {
+        private T value;
+        private Node prev;
+        private Node next;
+
+        public Node(T value, Node prev, Node next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node head;
+    private Node tail;
+
+    private int listSize;
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node(value, null, null);
+            tail = head;
+        } else {
+            Node newNode = new Node(value, tail, null);
+            tail.next = newNode;
+            tail = newNode;
+        }
+        listSize++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == listSize) {
+            add(value);
+            return;
+        }
+        checkIndexIsValid(index);
+        if (index == 0) {
+            head = new Node(value, null, head);
+            head.next.prev = head;
+        } else {
+            Node newNode = new Node(value, getNodeByIndex(index).prev, getNodeByIndex(index));
+            newNode.prev.next = newNode;
+            newNode.next.prev = newNode;
+        }
+        listSize++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T el : list) {
+            add(el);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node nodeToSet = getNodeByIndex(index);
+        T old = nodeToSet.value;
+        nodeToSet.value = value;
+        return old;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node nodeToRemove = getNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        while (currentNode != null) {
+            if (Objects.equals(currentNode.value, object)) {
```

Comment:
i think we still can not use Objects.equals and need to write implementetion yourself, but i am not sure